File: src/main/java/org/raku/annotation/fix/IdiomaticLoopFix.java
package org.raku.annotation.fix;

import com.intellij.codeInsight.intention.IntentionAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.util.IncorrectOperationException;
import org.raku.psi.RakuElementFactory;
import org.raku.psi.RakuWhileStatement;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

public class IdiomaticLoopFix implements IntentionAction {
    private final RakuWhileStatement myWhileStatement;

    public IdiomaticLoopFix(RakuWhileStatement whileStatement) {
        this.myWhileStatement = whileStatement;
    }

    @Nls
    @NotNull
    @Override
    public String getText() {
        return "Use idiomatic 'loop' syntax";
    }

    @Nls
    @NotNull
    @Override
    public String getFamilyName() {
        return getText();
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
        return true;
    }

    @Override
    public void invoke(@NotNull Project project, Editor editor, PsiFile file) throws IncorrectOperationException {
        PsiElement loop = RakuElementFactory.createLoop(project, myWhileStatement.getBlock());
        myWhileStatement.replace(loop);
    }

    @Override
    public boolean startInWriteAction() {
        return true;
    }
}

File: src/main/java/org/raku/comma/inspection/fixes/IdiomaticLoopFix.kt
package org.raku.comma.inspection.fixes

import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import org.raku.comma.psi.RakuElementFactory
import org.raku.comma.psi.RakuWhileStatement

class IdiomaticLoopFix : LocalQuickFix {
    override fun getFamilyName(): String { return "Use idiomatic 'loop' syntax" }

    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
        val whileStatement = descriptor.psiElement as? RakuWhileStatement ?: return
        val loop: PsiElement = RakuElementFactory.createLoop(project, whileStatement.getBlock())
        whileStatement.replace(loop)
    }
}